CIP-0020/CIP-0020.md

```diff
@@ -0,0 +1,105 @@
+---
+CIP: 0016
+Title: Canonical CBOR serialisation
+Authors: Gabriel Kerekes <gabriel.kerekes@vacuumlabs.com>, Rafael Korbas <rafael.korbas@vacuumlabs.com>
+Comments-URI: https://forum.cardano.org/t/stake-pool-cold-keys-hd-derivation/43360
+Status: Draft
+Type: Standards
+Created: 2021-06-14
+License: CC-BY-4.0
+---
+
+## Abstract
+
+This CIP defines a canonical format for CBOR serialisation in Cardano by extending the suggestions for canonical CBOR from the [CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9).
+
+## Motivation
+
+It’s desirable to have interoperability between different transaction signing tools (e.g. cardano-cli, HW wallets, cardano-hw-cli), some of which may need to serialise the transaction by itselves (especially HW wallets, due to security/performance considerations). By interoperability we mean that given the same inputs (transaction parameters) the tools would produce the same outputs (e.g. transaction hash).
+
+This is particularly relevant with the introduction of multisig. If the party initially creating the transaction uses e.g. cardano-cli to create the transaction, then other involved multisig parties might not be able to sign the transaction using HW wallets, because HW wallets might order some elements in the transaction differently thus they would come up with a different transaction hash. The same applies when the transaction is initially created by a HW wallet and requires signatures from parties using cardano-cli.
+
+For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the signature returned would thus be invalid for the transaction created by cardano-cli.
```

Comment:
```suggestion
For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the returned witness signature would thus be invalid for the transaction created by cardano-cli.
```